go/parquet/schema/logical_types.go

```diff
@@ -0,0 +1,1089 @@
+	return ConvertedTypes.JSON, DecimalMetadata{}
+}
+
+func (JSONLogicalType) IsCompatible(c ConvertedType, dec DecimalMetadata) bool {
+	return c == ConvertedTypes.JSON && !dec.IsSet
+}
+
+func (JSONLogicalType) IsApplicable(t parquet.Type, _ int32) bool {
+	return t == parquet.Types.ByteArray
+}
+
+func (JSONLogicalType) toThrift() *format.LogicalType {
+	return &format.LogicalType{JSON: format.NewJsonType()}
+}
+
+func (JSONLogicalType) Equals(rhs LogicalType) bool {
+	_, ok := rhs.(JSONLogicalType)
+	return ok
+}
+
+// BSONLogicalType represents a binary JSON string in the byte array
+type BSONLogicalType struct{ baseLogicalType }
+
+func (BSONLogicalType) SortOrder() SortOrder {
+	return SortUNSIGNED
+}
+
+func (BSONLogicalType) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]string{"Type": BSONLogicalType{}.String()})
+}
+
+func (BSONLogicalType) String() string {
+	return "BSON"
+}
+
+func (BSONLogicalType) ToConvertedType() (ConvertedType, DecimalMetadata) {
+	return ConvertedTypes.BSON, DecimalMetadata{}
+}
+
+func (BSONLogicalType) IsCompatible(c ConvertedType, dec DecimalMetadata) bool {
+	return c == ConvertedTypes.BSON && !dec.IsSet
+}
+
+func (BSONLogicalType) IsApplicable(t parquet.Type, _ int32) bool {
+	return t == parquet.Types.ByteArray
+}
+
+func (BSONLogicalType) toThrift() *format.LogicalType {
+	return &format.LogicalType{BSON: format.NewBsonType()}
+}
+
+func (BSONLogicalType) Equals(rhs LogicalType) bool {
+	_, ok := rhs.(BSONLogicalType)
+	return ok
+}
+
+// UUIDLogicalType can only be used with a FixedLength byte array column
+// that is exactly 16 bytes long
+type UUIDLogicalType struct{ baseLogicalType }
+
+func (UUIDLogicalType) SortOrder() SortOrder {
+	return SortUNSIGNED
+}
+
+func (UUIDLogicalType) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]string{"Type": UUIDLogicalType{}.String()})
+}
+
+func (UUIDLogicalType) String() string {
+	return "UUID"
+}
+
+func (UUIDLogicalType) ToConvertedType() (ConvertedType, DecimalMetadata) {
+	return ConvertedTypes.None, DecimalMetadata{}
+}
+
+func (UUIDLogicalType) IsCompatible(c ConvertedType, dec DecimalMetadata) bool {
+	if dec.IsSet {
+		return false
+	}
+	switch c {
+	case ConvertedTypes.None, ConvertedTypes.NA:
+		return true
+	}
+	return false
+}
+
+func (UUIDLogicalType) IsApplicable(t parquet.Type, tlen int32) bool {
+	return t == parquet.Types.FixedLenByteArray && tlen == 16
+}
+
+func (UUIDLogicalType) toThrift() *format.LogicalType {
+	return &format.LogicalType{UUID: format.NewUUIDType()}
+}
+
+func (UUIDLogicalType) Equals(rhs LogicalType) bool {
+	_, ok := rhs.(UUIDLogicalType)
+	return ok
+}
+
+// IntervalLogicalType is not yet in the thrift spec, but represents
+// an interval time and needs to be a fixed length byte array of 12 bytes
+type IntervalLogicalType struct{ baseLogicalType }
+
+func (IntervalLogicalType) SortOrder() SortOrder {
+	return SortUNKNOWN
```

Comment:
i believe the parquet spec has this sort order unsigned (which doesn't really make sense to me)

Reply:
https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#interval

> The sort order used for INTERVAL is undefined. When writing data, no min/max statistics should be saved for this type and if such non-compliant statistics are found during reading, they must be ignored.

😸

Reply:
Looks like a doc update is in order [parquet.thrift](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L875) seems to state otherwise.

Reply:
are you saying that the comment in parquet.thrift needs to be updated? Or the docs I pointed at need to be updated and I should change this? Sorry, it's just unclear from your comment and I'm not sure if parquet.thrift is considered the source of truth or if the parquet-format repo is lol

Reply:
I agree with the docs you linked to, so I think parquet.thrift is wrong in this case.